components/browser/state/src/main/java/mozilla/components/browser/state/action/BrowserAction.kt

```diff
@@ -152,6 +152,20 @@ sealed class TabListAction : BrowserAction() {
      */
     data class AddMultipleTabsAction(val tabs: List<TabSessionState>) : TabListAction()
 
+    /**
+     * Moves a set of tabIDs into a new position (maintaining internal ordering)
+     *
+     * @property tabIds The IDs of the tabs to move.
+     * @property position The position in the list to move them to.
+     * @property tabsFilter Apply a filter to the tabs to offset the [position].
+     * Needed to deal with the inactive tabs not being in the main tab tray.
+     */
+    data class MoveTabsAction(
+        val tabIds: List<String>,
+        val position: Int,
+        val tabsFilter: (TabSessionState) -> Boolean = { true }
+    ) : TabListAction()
```

Comment:
Can you elaborate a bit why this specific API is the one you came up with?

For manual sorting (done by the user) I'd expect just one tab getting moved to a position. For only visual reordering (e.g. by "last access") we may prefer to do this at the UI layer. In other cases we may want to specify the way we want to order and order inside the reducer (instead of ordering outside and the passing the result along as an action).

Reply:
This is intended for manual sorting, yes. In Fenix the simplest interface (long-touch a tab to begin dragging it) is in conflict with the already-existing interface where long-touching a tab will enter multi-select mode. Since long-touching while in multi-select mode doesn't do anything right now, that was the cleanest place to support dragging. Since we'd already be in the multi-select mode, I figured it made sense to move all of the selected tabs.

In addition, it's trivial to use this interface to move a single tab (just give a collection of 1) whereas it's quite difficult to move multiple tabs at once using an interface that only moves one tab at a time (assuming you care about the positioning of the tabs relative to each other).

I do agree that for automated reordering it makes sense to do the sorting somewhere else. I'm just trying to get feature parity with Fennec (for the features I use) since I really want to get off of such an old/vulnerable browser.

Reply:
Oh and the tab filter is needed since the position is the post-filtering position, which we need to convert to a pre-filtered position and the filter used is necessary to do that.